### PR TITLE
Change config format to scope configs by type

### PR DESCRIPTION
**Pull request by hdevalence**
Closes: #3636

## Description

This turns out to be a very good place to start to discover requirements for Hermes to operate in a more multi-chain way, since it allows discovering which parts of the code access Cosmos-specific data to do Cosmos-specific logic, and which parts are required for general relaying.

The only change to the existing config format so far is the change in semantics of the `type` field, which now must be `CosmosSdk`. I did not preserve the existing logic and tests that parsing different variations of casing and dashes are supported, because I don't think this is particularly valuable to support.

So far, only the core `ibc-relayer` crate is updated; the cli is not, because it's not necessarily clear how the cli should change to accomodate non-SDK chains.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer/src/chain.rs
Comment (hdevalence) [inline crates/relayer/src/chain.rs:21]: This is gone entirely. Rather than having the config store an enum _of_ chain types, the config _is_ an enum, so that it can have different subtypes for different configs.

File: crates/relayer/src/chain/client.rs
Comment (hdevalence) [inline crates/relayer/src/chain/client.rs:29]: It seems like the right move here is to split out the Tendermint-related fields on the config into a Tendermint-specific config structure. Then that structure can be a field on both `CosmosSdkConfig` and other Tendermint-using chains (like Penumbra or Namada).

File: crates/relayer/src/chain/cosmos.rs
Comment (hdevalence) [inline crates/relayer/src/chain/cosmos.rs:880]: (Currently, the pattern is irrefutable, but it won't be in the future, and that should be handled now)

File: crates/relayer/src/chain/cosmos.rs
Comment (hdevalence) [inline crates/relayer/src/chain/cosmos.rs:955]: cf #3641

File: crates/relayer/src/config.rs
Comment (hdevalence) [inline crates/relayer/src/config.rs:34]: oops, will clean up after next round of discussion

File: crates/relayer/src/config.rs
Comment (hdevalence) [inline crates/relayer/src/config.rs:157]: ditto above cleanup comment

File: crates/relayer/src/config.rs
Comment (hdevalence) [inline crates/relayer/src/config.rs:564]: As mentioned in the issue description for #3636, this will produce TOML like
```toml
[[chains]]
type = "CosmosSdk"
id = "chain_id"
rpc_addr = "http://127.0.0.1:26657/"
grpc_addr = "http://127.0.0.1:8080/"
```
The only difference from the current format is that the `type` field is now mandatory, and it must be `CosmosSdk`, not `cosmos-sdk` or `cOsMo-sSdk` or any other variation. I don't think that supporting arbitrary variations is a worthwhile complexity vs ergonomics tradeoff.

File: crates/relayer/src/config.rs
Comment (hdevalence) [inline crates/relayer/src/config.rs:584]: These are the attributes that are used by Hermes outside of the Cosmos backend.

All of the other attributes are only used inside the Cosmos backend, so this is effectively a way to discover which parts of the configuration data is used by the relayer engine, and which parts are used only by the chain backend.

All other backends' configuration objects will need these fields.

An alternative representation would be to have the `ChainConfig` contain these common fields unconditionally, AND an inner enum with only the non-common configuration data. But I think this is actually worse, because now the backends can't cleanly model their own data: they need access to common fields like the chain id, and those fields would live in an overarching `ChainConfig` object. With the current structure, the common fields would be duplicated across different objects (the `PenumbraConfig` would have an `id` field as well), but this actually makes the code more modular, because each backend can just pass around and work with one object it defines (the cosmos logic can work with a `CosmosSdkConfig`, the penumbra logic can work with a `PenumbraConfig`, and so on).

File: crates/relayer/src/keyring.rs
Comment (hdevalence) [inline crates/relayer/src/keyring.rs:290]: Is there any reason this isn't a method on `ChainConfig`, rather than a bare function defined in this other module?

**Comment by hdevalence** [comment]
I was planning to leave self-review comments on the build failures that reveal issues with the CLI interface, but that will need to wait for the CI to run.

File: crates/relayer/src/config.rs
Comment (hdevalence) [inline crates/relayer/src/config.rs:588]: This should move to the `cosmos/` backend but I left it for now for ease of initial commentary/review.

**Comment by hdevalence** [comment]
I thought it'd be possible to add comments to check failures, but this doesn't seem to be the case. Instead I put them inline.

File: crates/relayer-cli/src/config.rs
Comment (hdevalence) [inline crates/relayer-cli/src/config.rs:101]: Shouldn't config validation code be attached to the configs in the core crate, rather than being in the cli? Happy to move it if that sounds good.

File: crates/relayer-cli/src/commands/clear.rs
Comment (hdevalence) [inline crates/relayer-cli/src/commands/clear.rs:82]: This ties in to https://github.com/informalsystems/hermes/issues/3641 

Perhaps the key manager should only support opaque byte strings, and let the rest be handled by the backend?

**Review by romac — COMMENTED** [review]
Thank you so much for taking on this whole endeavor. What you've done so far looks really good to me.

Now we need to figure out a good story for backend-specific commands, eg. `keys`, as you've pointed out here and in #3641.

**Comment by romac** [comment]
> Now we need to figure out a good story for backend-specific commands, eg. `keys`, as you've pointed out here and in #3641.

Maybe for now we can restrict these commands to `CosmosSdk` chains, by throwing a runtime error if used against another type of chain?

**Comment by erwanor** [comment]
I have done some work propagating the configuration changes, but I'm leaving some of the refactor items as a last step (as previously agreed above).

I would like to pause and see if CI passes, tie up loose ends, before pushing on the specific TODOs (e.g. extracting the tendermint config) and doing another pass.

**Comment by erwanor** [comment]
This took longer than anticipated, but the branch is finally passing all integration tests. Now, we should be able to move quickly on getting the remaining items done and ready for review by tomorrow:
- [x] small review issues
    - [x] remove dead code
    - [x] move `list_keys` to be a method on `ChainConfig`
    - [x] move config validation out of `relayer-cli`
 
when this is done:
- [ ] extract the tendermint fields into a struct and make the chain config embed it

and finally:
- [ ] move `CosmosSdkConfig` into the `cosmos/` backend
- [ ] move the cosmos specific config validation routines into the sdk backend
- [ ] move `detect_compatibility_mode` into the cosmos sdk impl

**Comment by erwanor** [comment]
@romac I think this is ready for review! We can split out the tendermint config in future work since it has a small surface area, ditto for the websocket methods, for now we can just error if they are not supported.

**Review by romac — APPROVED** [review]
Amazing work 🚀 Thanks so much to both of you for seeing this through, much appreciated! 💐

**Comment by romac** [comment]
@hdevalence Before I can merge this, can you please enable edits from maintainers on this PR?

**Comment by hdevalence** [comment]
Hmm, the Github docs only talk about personal-owned forks, I'm not sure if it's possible because we made it from an org fork. But I can certainly add you to our org fork ? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#

**Comment by erwanor** [comment]
@romac I can't see that button, maybe it's only visible to the PR creator, but I added you as a collaborator to the forked repository: [invitations](https://github.com/penumbra-zone/hermes/invitations)

**Comment by romac** [comment]
Perfect, thank you both! 🙏

**Comment by seanchen1991** [comment]
Will these config changes require updates to the Hermes guide?

**Comment by romac** [comment]
I thought they were all taken care of, but actually found three templates that we missed: https://github.com/informalsystems/hermes/pull/3679

**Pull request merged**
